# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/ClientWebGLContext.h
# Commit: 7b82837e2f56
# Full Hash: 7b82837e2f56582e1e86897285f4109856b72000
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2021-11-11 15:25:44
# Regressor Bug: 1739448
# File Overlap Count: 1
# Description:
#   Bug 1739448 - Implement a prototype WebGL-backed Canvas2D. r=gfx-reviewers,nical,jrmuizel
#   
#   This mainly provides DrawTargetWebgl, which implements the subset of the DrawTarget
#   API necessary for integration with CanvasRenderingContext2D. It translates them to
#   suitable commands for its internal ClientWebGLContext, which then manages remoting
# ==============================================================================

diff -r feac99f1dae3 -r 7b82837e2f56 dom/canvas/ClientWebGLContext.h
--- a/dom/canvas/ClientWebGLContext.h	Thu Nov 11 04:33:50 2021 +0000
+++ b/dom/canvas/ClientWebGLContext.h	Thu Nov 11 04:33:50 2021 +0000
@@ -37,6 +37,10 @@
 class WebGLChild;
 }
 
+namespace gfx {
+class DrawTargetWebgl;
+}
+
 namespace webgl {
 class AvailabilityRunnable;
 class TexUnpackBlob;
@@ -693,6 +697,7 @@
   friend class webgl::ObjectJS;
   friend class webgl::ProgramKeepAlive;
   friend class webgl::ShaderKeepAlive;
+  friend class gfx::DrawTargetWebgl;
 
   // ----------------------------- Lifetime and DOM ---------------------------
  public:
@@ -921,6 +926,12 @@
   }
 
   void OnMemoryPressure() override;
+  void SetContextOptions(const WebGLContextOptions& aOptions) {
+    mInitialOptions.emplace(aOptions);
+  }
+  const WebGLContextOptions& GetContextOptions() const {
+    return mInitialOptions.ref();
+  }
   NS_IMETHOD
   SetContextOptions(JSContext* cx, JS::Handle<JS::Value> options,
                     ErrorResult& aRvForDictionaryInit) override;
@@ -1385,6 +1396,8 @@
   void BufferData(GLenum target, const dom::ArrayBufferView& srcData,
                   GLenum usage, GLuint srcElemOffset = 0,
                   GLuint srcElemCountOverride = 0);
+  void RawBufferData(GLenum target, const Range<const uint8_t>& srcData,
+                     GLenum usage);
 
   void BufferSubData(GLenum target, WebGLsizeiptr dstByteOffset,
                      const dom::ArrayBufferView& src, GLuint srcElemOffset = 0,
@@ -1503,6 +1516,9 @@
                   GLenum internalFormat, const ivec3& size) const;
 
   // Primitive tex upload functions
+  void RawTexImage(uint32_t level, GLenum respecFormat, uvec3 offset,
+                   const webgl::PackingInfo& pi,
+                   const webgl::TexUnpackBlobDesc&) const;
   void TexImage(uint8_t funcDims, GLenum target, GLint level,
                 GLenum respecFormat, const ivec3& offset, const ivec3& size,
                 GLint border, const webgl::PackingInfo& pi,